### PR TITLE
feat(datepicker): ng-model-options: allowInvalid support

### PR DESCRIPTION
**Pull request by jbarrus**
Relevant issues - #4694, #4837 

Somewhat based on https://github.com/davious/angular-ui-bootstrap/commit/cba7a614e37cdc34a56eaac8716af3d985b99b9d from @davious

### CODE REVIEW

File: src/datepicker/datepicker.js [hunk 1]
```diff
@@ -605,6 +611,9 @@ function(scope, element, attrs, $compile, $parse, $document, $rootScope, $positi
     datepickerEl = angular.element(popupEl.children()[0]);
     datepickerEl.attr('template-url', datepickerTemplateUrl);
 
+    scope.ngModelOptions = { $options: angular.extend({ allowInvalid: false }, ngModelOptions) };
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Is there any particular reason that this should be exposed on scope?

File: src/datepicker/datepicker.js [hunk 2]
```diff
@@ -543,19 +547,21 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
...
-    ngModel, $popup, altInputFormats;
+    ngModel, ngModelOptions, $popup, altInputFormats;
 
   scope.watchData = {};
 
-  this.init = function(_ngModel_) {
+  this.init = function(_ngModel_, _ngModelOptions_) {
     ngModel = _ngModel_;
+
+    ngModelOptions = _ngModelOptions_ && _ngModelOptions_.$options || {};
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: I believe one can pull off the options from `ngModel.$options`

File: src/datepicker/datepicker.js (hunk 1 above)
Comment (jbarrus) [inline src/datepicker/datepicker.js]: I don't think it's necessary, I copied some of the code from @davious (forgot to mention that originally in the pull request) and he had it in there.

File: src/datepicker/datepicker.js (hunk 2 above)
Comment (jbarrus) [inline src/datepicker/datepicker.js]: Do you mean it's safe to access without checking?

Comment (davious) [inline src/datepicker/datepicker.js]: @jbarrus @wesleycho is right. All that ngModelOption controller hook-up seems unnecessary; my bad. We can just look at `ngModel.$options`. I plan to update my wip soon.

**Comment by jbarrus** [comment]
I see, thanks, ngModel.$options is definitely much simpler.  Does this look right?

**Comment by wesleycho** [comment]
Oops...I messed up with my commit - this is now in as 2460e423224492a35a354ee6f3dc0dd4352d00d6

**Pull request closed without merging**